### PR TITLE
HPCC-17757 Couchbase Plugin: Improve parsing of complex query result

### PR DESCRIPTION
**Pull request by dcamper**
Signed-off-by: Dan S. Camper <dan.camper@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:

Tested with various queries against a Couchbase 4.6 instance.  Query results were of varying complexity (depth of ECL records, inclusion of child records and child datasets, etc.) including the example in the JIRA ticket.

### CODE REVIEW

**Comment by dcamper** [comment]
@rpastrana Please review.

File: plugins/couchbase/couchbaseembed.cpp [hunk 1]
```diff
@@ -899,68 +899,145 @@ namespace couchbaseembed
...
+        else
+        {
+            xpath = str(field->name);
+        }
+
+        return xpath;
+    }
 
-        if (!m_oResultRow->hasProp(fieldname))
+    void CouchbaseRowBuilder::constructNewXPath(StringBuffer& outXPath, const char * nextNode = NULL) const
```
Comment (rpastrana) [inline plugins/couchbase/couchbaseembed.cpp]: arg default should be in declaration

File: plugins/couchbase/couchbaseembed.cpp [hunk 2]
```diff
@@ -899,68 +899,145 @@ namespace couchbaseembed
...
+    {
+        for (std::vector<PathTracker>::const_iterator x = m_pathStack.begin(); x != m_pathStack.end(); x++)
```
Comment (rpastrana) [inline plugins/couchbase/couchbaseembed.cpp]: Minor, but consider renaming x to a more meaningful name

File: plugins/couchbase/couchbaseembed.cpp [hunk 3]
```diff
@@ -899,68 +899,145 @@ namespace couchbaseembed
...
+        const char * xpath = xpathOrName(field);
 
-        if (getNumFields(field->type->queryChildType()) > 0)
-            m_oNestedField.set(m_oResultRow->queryBranch(field->name->queryStr()));
+        if (xpath && *xpath)
+        {
+            PathTracker     newPathNode(xpath, true);
+            StringBuffer    newXPath;
+
+            constructNewXPath(newXPath, xpath);
```
Comment (rpastrana) [inline plugins/couchbase/couchbaseembed.cpp]: Is it necessary to keep full XPATH?

**Review by rpastrana — COMMENTED** [review]
@dcamper a few comments

File: plugins/couchbase/couchbaseembed.cpp [hunk 4]
```diff
@@ -899,68 +899,145 @@ namespace couchbaseembed
...
-         *parentRec := RECORD
-         * childRec child1,                        <-- flatens out the childrec, this function would receive a field of name x
-         * dataset(childRec) child2;               <-- keeps nested structure, this funciton would receive a field of name child2
-         *END;
-        */
+        const char * xpath = xpathOrName(field);
 
-        if (getNumFields(field->type->queryChildType()) > 0)
-            m_oNestedField.set(m_oResultRow->queryBranch(field->name->queryStr()));
+        if (xpath && *xpath)
```
Comment (dcamper) [inline plugins/couchbase/couchbaseembed.cpp]: Yes, here and in other places where the check was made but no error thrown. Fixed here and in those places. Also adjusted error message to give a better idea of what was happening during parse when the error appeared.

File: plugins/couchbase/couchbaseembed.cpp (hunk 3 above)
Comment (dcamper) [inline plugins/couchbase/couchbaseembed.cpp]: Unfortunately, yes. XPATHs can be relative (which are more common) or absolute, or at least "full" in the sense that they must start processing at the root of the property tree. This is a topic that may bear revisiting at some point in the future, though.

File: plugins/couchbase/couchbaseembed.cpp (hunk 1 above)
Comment (dcamper) [inline plugins/couchbase/couchbaseembed.cpp]: Copy and paste error which shouldn't have been there to begin with; removed.

File: plugins/couchbase/couchbaseembed.cpp (hunk 2 above)
Comment (dcamper) [inline plugins/couchbase/couchbaseembed.cpp]: Renamed to 'iter'.

**Comment by dcamper** [comment]
@rpastrana Back to you.

**Comment by rpastrana** [comment]
@dcamper looks good.

**Comment by richardkchapman** [comment]
@dcamper Please squash to a single commit with a properly formatted title line - this should start with the JIRA number and make sense as a single line comment in the change log. Something like:

HPCC-17757 Couchbase Plugin: Improve parsing of complex query result

would be perfect.

**Comment by dcamper** [comment]
@richardkchapman Done, and thanks for the reminder to format the commits and not the PR.

**Pull request merged**